### ISSUE TITLE
Can DALI be integrated into HuggingFace Trainer?

### ISSUE THREAD

**Issue by ShyFoo** [NVIDIA/DALI#5421]
### Describe the question.

Is it possible to use DALI as the dataloader for HuggingFace Trainer?

### Check for duplicates

- [X] I have searched the [open bugs/issues](https://github.com/NVIDIA/DALI/issues) and have found no duplicates for this bug report

**Comment by JanuszL** [comment]
Hi ShyFoo,

Thank you for reaching out. While we haven't tried this directly yet DALI provides compatibility with [PyTorch](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/plugins/pytorch_tutorials.html) so you can create a DALI iterator that would behave as PyTorch one.
Have you tried this yet? Have you encountered any issues?

**Comment by JanuszL** [comment]
I just checked and unfortunately, HuggingFace trainer (I checked only one example so far) expects PyTorch DataIterator (checks for the type of passes object). I still believe it is possible to make DALI work there but I may need a bit of adjustments to the internal methods. If you have some spare time we would be more than happy to review a PR with such an example.

**Comment by ShyFoo** [comment]
> I just checked and unfortunately, HuggingFace trainer (I checked only one example so far) expects PyTorch DataIterator (checks for the type of passes object). I still believe it is possible to make DALI work there but I may need a bit of adjustments to the internal methods. If you have some spare time we would be more than happy to review a PR with such an example.

Yeah, I have tried integrating DALI into the HuggingFace trainer. As you mentioned, it seems to expect either _torch.utils.data.Dataset_ or _torch.utils.data.IterableDataset_ as input. It might be possible to customize a data collator, which can preprocess data in a DALI pipeline, for the HuggingFace trainer, but I'm not sure if it will work. If you have any idea for this, please let me know. I'm sure it would be a huge step for the whole community!

**Comment by JanuszL** [comment]
I did some basic investigation:
- for CV tasks DALI needs to have such wrapper:
  ```
  class DALIwrapper(DALIGenericIterator):
      def __init__(self, *args, **kargs):
          super().__init__(*args, **kargs)
  
      def __next__(self):
          batch = super().__next__()
          return batch[0]
      
  pipe_train = GetDaliPipeline(...)
  data_train = DALIwrapper([pipe_train], ["pixel_values", "labels"], reader_name="Reader")
  ```
- trainer needs to have these methods replaced:
  ```
  trainer.get_train_dataloader = lambda : data_train
  trainer.get_eval_dataloader = lambda x: x
  ```
  so DALI dataloader is used as is without wrapping into PyTorch DataIterator
- there is probably another layer of complexity using DALI with a distributed strategy which is handled by accelerate

**Comment by ShyFoo** [comment]
For the third point, _get_train_dataloader_ returns _self.accelerator.prepare(DataLoader(train_dataset, **dataloader_params))_ for a distributed training or evaluation. So what should I do in this setting? Hope for your help.

**Comment by JanuszL** [comment]
@ShyFoo - I'm afraid I don't know the answer. My idea was to avoid wrapping DALI into DataLoader as it serves as one already. If you have time I would be more than happy to learn your findings.

**Comment by ShyFoo** [comment]
> e basic investigation:

Got it. I have tried using DALI + DeepSpeed, where I have to define a triaing loop and other practical functions myself, instead of using the Huggingface Trainer for convience, so I prefer to use DALI, DeepSpeed, and Huggingface Trainer simultaneously if possible.

**Comment by JanuszL** [comment]
If you have any examples of DALI + DeepSpeed feel free to post them as PR to DALI - we would be more than happy to enrich our documentation and the base of samples.

**Comment by ShyFoo** [comment]
Well, I'd be happy to.

**Comment by Ace1169** [comment]
I was thinking this might work but haven't had a chance to try yet....been swamped with other development: But like the idea and a min to write out what I have been kicking around the brain bucket! 

Adding DALI to the Mix
To add DALI, you'll need to create a custom dataset class that bridges DALI pipelines with the Huggingface ecosystem:

import nvidia.dali as dali
from nvidia.dali.plugin.pytorch import DALIGenericIterator
from torch.utils.data import Dataset

class DALIDataset(Dataset):
    def __init__(self, pipeline, size):
        self.pipeline = pipeline
        self.size = size
        self.iterator = DALIGenericIterator([pipeline], ["data", "label"], size)
        
    def __len__(self):
        return self.size
    
    def __getitem__(self, idx):
        if self.iterator._counter >= self.size:
            self.iterator.reset()
        data = next(self.iterator)[0]
        # Convert to Huggingface format here
        return {
            "input_ids": data["data"],
            "labels": data["label"]
            # Add other required fields based on your model needs
        }

Setting It All Up
# 1. Create your DALI pipeline
pipeline = create_dali_pipeline()  # Your function to create and return a DALI pipeline

# 2. Wrap it in the custom dataset
dali_dataset = DALIDataset(pipeline, size=10000)  # Or however many samples you have

# 3. Configure DeepSpeed
# Create ds_config.json (can be done programmatically too)

# 4. Create Trainer with DeepSpeed and DALI dataset
trainer = Trainer(
    model=model,
    args=TrainingArguments(
        output_dir="./output",
        deepspeed="ds_config.json",
        # Other args
    ),
    train_dataset=dali_dataset,
)

# 5. Train!
trainer.train()

**Comment by ShyFoo** [comment]
> I was thinking this might work but haven't had a chance to try yet....been swamped with other development: But like the idea and a min to write out what I have been kicking around the brain bucket!
> 
> Adding DALI to the Mix To add DALI, you'll need to create a custom dataset class that bridges DALI pipelines with the Huggingface ecosystem:
> 
> import nvidia.dali as dali from nvidia.dali.plugin.pytorch import DALIGenericIterator from torch.utils.data import Dataset
> 
> class DALIDataset(Dataset): def **init**(self, pipeline, size): self.pipeline = pipeline self.size = size self.iterator = DALIGenericIterator([pipeline], ["data", "label"], size)
> 
> ```
> def __len__(self):
>     return self.size
> 
> def __getitem__(self, idx):
>     if self.iterator._counter >= self.size:
>         self.iterator.reset()
>     data = next(self.iterator)[0]
>     # Convert to Huggingface format here
>     return {
>         "input_ids": data["data"],
>         "labels": data["label"]
>         # Add other required fields based on your model needs
>     }
> ```
> 
> Setting It All Up
> 
> # 1. Create your DALI pipeline
> pipeline = create_dali_pipeline() # Your function to create and return a DALI pipeline
> 
> # 2. Wrap it in the custom dataset
> dali_dataset = DALIDataset(pipeline, size=10000) # Or however many samples you have
> 
> # 3. Configure DeepSpeed
> # Create ds_config.json (can be done programmatically too)
> # 4. Create Trainer with DeepSpeed and DALI dataset
> trainer = Trainer( model=model, args=TrainingArguments( output_dir="./output", deepspeed="ds_config.json", # Other args ), train_dataset=dali_dataset, )
> 
> # 5. Train!
> trainer.train()

Thank you very much! I will try your provided solution.
